akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -124,19 +124,33 @@ private[akka] final case class Recover[T](pf: PartialFunction[Throwable, T]) ext
 /**
  * INTERNAL API
  */
-private[akka] final case class Take[T](count: Long) extends PushPullStage[T, T] {
-  private var left: Long = count
+private[akka] final case class Take[T](count: Long) extends SimpleLinearGraphStage[T] {
+  override def initialAttributes: Attributes = DefaultAttributes.take
 
-  override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-    left -= 1
-    if (left > 0) ctx.push(elem)
-    else if (left == 0) ctx.pushAndFinish(elem)
-    else ctx.finish() //Handle negative take counts
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+    private var left: Long = count
+
+    override def onPush(): Unit = {
+      left -= 1
+      if (left > 0) {
+        push(out, grab(in))
+      } else if (left == 0) {
+        push(out, grab(in))
+        completeStage()
+      } else {
+        completeStage()
+      }
+    }
+
```

Comment:
@trautonen I'd probably do something like this:

~~~scala
    override def onPush(): Unit = {
      val leftBefore = left
      left = leftBefore - 1
      if (leftBefore >= 1) push(out, grab(in))
      if (leftBefore <= 1) completeStage()
    }
~~~

Why? Because I prefer code to be explicit that it doesn't depend on other logic manipulating the `var` somewhere in the callchain of the method. :)